### PR TITLE
[MOB-16019] Set pending state before updating IdentityMap

### PR DESCRIPTION
**Pull request by addb**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: Sources/IdentityState.swift [hunk 1]
```diff
@@ -104,8 +104,8 @@ class IdentityState {
     ///
     /// - Parameters
     ///   - event: event containing customer identifiers to add or update with the current customer identifiers
-    ///   - createXDMSharedState: function which creates new XDM shared state
-    func updateCustomerIdentifiers(event: Event, createXDMSharedState: ([String: Any], Event) -> Void) {
+    ///   - resolveXDMSharedState: function which resolves pending XDM shared state
+    func updateCustomerIdentifiers(event: Event, resolveXDMSharedState: ([String: Any]) -> Void) {
         guard let identifiersData = event.data else {
             Log.debug(label: IdentityConstants.FRIENDLY_NAME, "IdentityState - Failed to update identifiers as no identifiers were found in the event data.")
             return
```
Comment (kevinlind) [inline Sources/IdentityState.swift:111]: The pending shared state needs to be resolved before returning from this function. At any point where this function can return, resolve the shared state by passing in the current Identity properties.

File: Sources/IdentityState.swift [hunk 2]
```diff
@@ -117,14 +117,14 @@ class IdentityState {
...
     /// Remove customer identifiers specified in `event` from the current `IdentityMap`.
     /// - Parameters:
     ///   - event: event containing customer identifiers to remove from the current customer identities
-    ///   - createXDMSharedState: function which creates new XDM shared states
-    func removeCustomerIdentifiers(event: Event, createXDMSharedState: ([String: Any], Event) -> Void) {
+    ///   - resolveXDMSharedState: function which resolves pending XDM shared states
+    func removeCustomerIdentifiers(event: Event, resolveXDMSharedState: ([String: Any]) -> Void) {
         guard let identifiersData = event.data else {
             Log.debug(label: IdentityConstants.LOG_TAG, "IdentityState - Failed to remove identifier as no identifiers were found in the event data.")
             return
```
Comment (kevinlind) [inline Sources/IdentityState.swift:130]: Same here and the next `return` statement, resolve the pending state with the current Identity properties.

File: Sources/IdentityState.swift (hunk 1 above)
Comment (kevinlind) [reply Sources/IdentityState.swift:111]: Add unit tests for these cases as well. Verify that the pending shared state is resolved after the update or remove customer identifiers methods are called, even when passed no or invalid identifiers to update/remove.

**Comment by kevinlind** [comment]
We should also add a pending state when `resetIdentifiers` is called as well as it updates the Identity Map and generates a new ECID.

File: Tests/UnitTests/IdentityStateTests.swift [hunk 3]
```diff
@@ -461,9 +462,9 @@ class IdentityStateTests: XCTestCase {
                           type: EventType.edgeIdentity,
```
Comment (emdobrin) [inline Tests/UnitTests/IdentityStateTests.swift:462]: please update the name of the test as well

File: Sources/Identity.swift [hunk 4]
```diff
@@ -64,20 +64,25 @@ import Foundation
...
+        // [MOB-16019] Adding pending shared state to avoid race condition between updating and reading identity map
+        let resolver = createPendingXDMSharedState(event: event)
+        state.updateCustomerIdentifiers(event: event, resolveXDMSharedState: resolver)
     }
 
     /// Handles remove identity requests to remove customer identifiers.
     /// - Parameter event: the identity request event
     private func handleRemoveIdentity(event: Event) {
-        state.removeCustomerIdentifiers(event: event, createXDMSharedState: createXDMSharedState(data:event:))
+        // [MOB-16019] Adding pending shared state to avoid race condition between updating and reading identity map
```
Comment (emdobrin) [inline Sources/Identity.swift:75]: let's remove the ticket id from this comment, it can be found based on commit history.

**Review by emdobrin — APPROVED** [review]

**Review by kevinlind — APPROVED** [review]

**Pull request merged**